Make gotthard work reliably when total number of DB stacks exceed 100.

alexeyklyukin opened the pull request:
Fix #11

alexeyklyukin committed:
Do not ignore subsequent chunks of CF stacks.

Fixes inability to find CF stacks for some cluster when total
number of stacks exceeds AWS limit for one chunk (100 by default).

alexeyklyukin committed:
Change references to 'mai' to 'zaws login'

alexeyklyukin committed:
PEP8 style fixes.

a1exsh commented on gotthard/cli.py:
```diff
@@ -307,13 +307,23 @@ def get_spilo_stacks(cf, spilo_cluster):
+    while True:
+        # AWS emits output in chunks, loop until we examine all of them.
+        current_chunk = cf.list_stacks(StackStatusFilter=stack_status) if not next_token \
+            else cf.list_stacks(StackStatusFilter=stack_status, NextToken=next_token)
```
I think what you actually want is this: http://boto3.readthedocs.io/en/latest/reference/services/cloudformation.html?highlight=%22PageIterator%22#CloudFormation.Paginator.ListStacks

Example use: https://github.com/zalando-zmon/zmon-aws-agent/blob/master/zmon_aws_agent/aws.py#L157  Though I don't see `build_full_result()` anywhere in the docs, I expect that to be supported across all paginator implementations.

avaczi commented on gotthard/cli.py:
```diff
@@ -224,12 +222,12 @@ def get_remote_host(remote_host, regions):
         if ce.response['Error']['Code'] in ['ExpiredToken', 'RequestExpired']:
             raise ClickException('\nAWS credentials have expired.\n' +
-                                 'Use the "mai" command line tool to get a new temporary access key.')
+                                 'Use the "zaws login" command line tool to get a new temporary access key.')
```
`zaws require`?

alexeyklyukin replied:
Good point, didn't know it exist, thanks!

avaczi commented on gotthard/cli.py:
```diff
@@ -245,7 +243,7 @@ def setup_tunnel(user, odd_host, remote_host, remote_port, tunnel_port):
                    '-N']
 
-    process = subprocess.Popen(ssh_command, preexec_fn = os.setpgrp)
+    process = subprocess.Popen(ssh_command, preexec_fn=os.setpgrp)
```
why this change? :D

alexeyklyukin replied:
PEP8: Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value.

avaczi replied:
bah.  Sounds very arbitrary :D

alexeyklyukin replied:
Why?

a1exsh commented:
I was a little surprised that we don't list boto in `requirements.txt`, had to browse source code in order to find the actual version being used.  It's `boto3`.

avaczi commented on gotthard/cli.py:
```diff
@@ -307,13 +305,23 @@ def get_spilo_stacks(cf, spilo_cluster):
-                if k == 'SpiloCluster' and (not spilo_cluster or v == spilo_cluster):
-                    yield stack_name
+    next_token = None
+    while True:
```
Isn't there a more 'sparing' (ie. more direct, without listing so many stacks) to find the stack name?  That would automatically remove the problem being solved here.

alexeyklyukin replied:
Not unless you hardcode the stack as spilo-{whatevername} (which might be not true for some other teams). 
There seem to be no way to partially match the name, see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html

alexeyklyukin replied:
Actually, I've checked the performance of the ``describe_stacks`` with the hard-coded name template of spilo-{version}, and it didn't show any significant speed improvements over ``list_stacks``. In cases of either getting through stacks manually or describing a specific stack the requests took from 4s to 9s and timed out occasionally for an unspecified reason.

avaczi replied:
I was trying to talk basically about if it is possible to filter the stack list _before_ receiving it, thus avoiding the whole paging issue and complicated code

alexeyklyukin replied:
That's exactly what describe_stacks do.

alexeyklyukin committed:
Address the code review

- show zaws require instead of zaws login in help messages
- use pagination for the list_stack with the reasonably large MaxItems.

alexeyklyukin committed:
add missing boto3 to the project requirements

alexeyklyukin committed:
remove spurious whitespaces.

a1exsh commented on gotthard/cli.py:
```diff
@@ -307,7 +305,11 @@ def get_spilo_stacks(cf, spilo_cluster):
-    for stack in cf.list_stacks(StackStatusFilter=stack_status)['StackSummaries']:
+    paginator = cf.get_paginator('list_stacks')
+    stacks = paginator.paginate(StackStatusFilter=stack_status,
+                                PaginationConfig = {'MaxItems': 10000}).build_full_result()['StackSummaries']
```
PEP8 :-p

a1exsh replied:
Why explicit MaxItems?  Do you really expect more than 10k CF stacks?

alexeyklyukin replied:
As far as I understood, MaxItems gives you the number of items you get on a single run, before you need to request the NextToken, so 10K is set to avoid calling this function multiple times.

a1exsh replied:
Why do you need the paginator then?  In case it's still over 10k?

alexeyklyukin replied:
On re-reading it, it's probably PageSize rather than MaxItems that I need.

alexeyklyukin replied:
Well, with Paginators I can explicitly set the page size to be more than 100 items. I cannot find an explicit guarantee there that there will be only one page, as build_full_result returns the NextToken in the source code and there is no documentation.

alexeyklyukin committed:
Tweak the paginator parameters.

Use the PageSize instead of the MaxItems. We want to set each page
to be large enough in practice to accomodate all stacks.

a1exsh commented on gotthard/cli.py:
```diff
@@ -307,7 +305,11 @@ def get_spilo_stacks(cf, spilo_cluster):
-    for stack in cf.list_stacks(StackStatusFilter=stack_status)['StackSummaries']:
+    paginator = cf.get_paginator('list_stacks')
+    stacks = paginator.paginate(StackStatusFilter=stack_status,
+                                PaginationConfig = {'PageSize': 10000}).build_full_result()['StackSummaries']
```
This still suffers from PEP8 problem...

alexeyklyukin committed:
Introduce spilo-stack-name option. Refactoring.

Hard-code spilo stacks as spilo-{version}, allow overriding it with
--spilo-stack-name option.

Use describe_stacks to avoid going through the list of all CF stacks.

a1exsh commented:
:+1:

alexeyklyukin commented:
👍
